### ISSUE TITLE
erroneous "is specified more than once, so this usage will be overwritten.(2783)"

### ISSUE THREAD

**Issue by clausreinke** [microsoft/TypeScript#39671]
**TypeScript Version:**  3.9.2, also in 4.0.0-beta (not in 3.8.3)

**Search Terms:**

2783

**Expected behavior:**

code should be accepted without error

**Actual behavior:**

error 2783 ist raised

**Related Issues:**
#38535,  #39113, #36779, #37740

**Code**
(the actual code builds `obj` in a loop, but this small example shows the issue; btw, if you delete the `x: 0,`, there is no error, which is also not correct)
```ts
let obj: {[key: string]: {x: number}} = {};
// erroneous type error:
//      'x' is specified more than once, so this usage will be overwritten.(2783)
//      input.tsx(3, 21): This spread always overwrites this property
obj['key'] = {x: 0, ...obj['key']}; // x would be in obj['key'], if 'key' was in obj
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
let obj = {};
// erroneous type error:
//      'x' is specified more than once, so this usage will be overwritten.(2783)
//      input.tsx(3, 21): This spread always overwrites this property
obj['key'] = Object.assign({ x: 0 }, obj['key']); // x would be in obj['key'], if 'key' was in obj

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "moduleResolution": 2,
    "target": "ES2017",
    "jsx": "React",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?ts=3.9.2#code/DYUwLgBA9gRgVgLggbwNoGsQE8kGcwBOAlgHYDmAuksgB5IkCuAtjCAQL7sQC8K7A3ACgA9MIhsCUEiCgNcEMFgAOIcQUkEEIsRF26A5DX0Qi83CoDGRAGZEQAEwhMoBVWAAWAQxLQSFkAA0ELhQCu6mEHKeZKoA7kTAwBCs0ABubLHEYGAgJAB0ABQATADsABwAzACU2np6pEoMYHlguDQFFUFFAIxVSAAq4WZKrp6OnsCxnljyUOkEmUQ58h4RI1AqBIqCsHCo+phY+hQ8KHQQAAxBeTe7+4fHAhCiEDQQsbLAjimk0PD32GOQRsEAOgPennkv12QA)
      

potentially relevant: https://stackoverflow.com/questions/47155141/spreading-undefined-in-array-vs-object

**Comment by MartinJohns** [comment]
Duplicate of #37688, #39113, #38589, and probably more. Used search terms: `overwritten`

**Comment by clausreinke** [comment]
yes, there are multiple related issues, including some apparently fixed ones; problem is, they usually have an additional quirk, which is often the part that gets fixed; 

I tried to reduce our issue to a small example with nothing extra, so it is not a duplicate of any of the listed issues. There may well be a common root to several of the related issues, and there are sufficiently many related issues that a collective tracking issue may be in order.

**Comment by RyanCavanaugh** [comment]
The real duplicate here is #13778 since the fix is to include `undefined` in the type of `obj['key']`

**Comment by clausreinke** [comment]
while that is a more interesting discussion (I'm only halfway through at this point) that clarifies some of the goals that led to the current situation, neither the fix nor the problem are quite as simple. 

In the real-world code, we reconstruct a tree from a flat database representation, and we know that the order in which the code accesses the tree while the tree is being build is correct. We just find it difficult to convince typescript's CFA of this fact.

I've extended the playground example a bit in the hope of illustrating the effect: [Playground Link](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=14&pc=11#code/C4TwDgpgBAyg9gWwgWQIZigXigbygbQGsIQAuKAZ2ACcBLAOwHMBdcvAD3PoFcEAjCNQA0UMlD5w4AGwip6UAL6KA3AChVAeg1QZwKHD4ArcvCRoM2HKt36j5AAqpqwWqikAeUynQA+LLlUoIKhUNihOKABGETEabmgFIUDgvjCIgCYY8gAzNwoE1SUtEIooR2dXDy9zHzV1AGM4eioobP8ACmIQSPIuuDbq9BEu9N6SfthEbzAASiw-K2DbQyISSOZ-DnIABhEAOgODFZGNhTUFdWz2gHJUa6FrvmuZoA)

- if we do not declare the map as Partial, typescript complains about overwriting
- if we do declare the map as Partial, typescript complains about missing fields

Is our only option to disable typechecking here?

**Comment by clausreinke** [comment]
okay, while thinking some more about #13778  and about how the example still doesn't quite represent our problem, I was able to come up with a way to explain to typescript what is happening. For others finding this issue:

- we were building up our map, so it starts out empty (Partial) and ends up full (not Partial) and we happen to know for each key, when it is either (otherwise we could not be sure our code works)
- so we were able to take the basic map type and coerce it as neccessary:
  - `Partial<MapType>` to begin with
  - `obj[key] as MapType[keytype]` when we are certain the key has been filled

slightly extended example: [Playground Link](https://www.typescriptlang.org/play/?ssl=4&ssc=1&pln=5&pc=1#code/C4TwDgpgBAyg9gWwgWQIZigXigbygbQGsIQAuKAZ2ACcBLAOwHMBdcvAD3PoFcEAjCNQA0UMlD5w4AGwip6UAL6KA3AChVM4FDh8AVuXhI0GbDlVQLUVGyicoARhFia3aAqHnLfG3YBMT8gAzVCkKN1UFNXUAYzh6KihArCgACmIQe3J0uCTDFHQRdN8skhzYRHywAEosAD5cTwsdXSISe2ZkjnIABhEAOgGU5taMjtQKKAAFVGpgWhCAHjzjWvwqOiZmKoCoFzc1S209Ed8O0ygBvqHjorGJ5fQ1mgYWbdFyPZUI9UCUgHJon8RH9UH8qqpfiCgVA-nwwTE4hRpBA+lI4IwUgApGAAeQAcn11i9aIEQNddCIeFIpCJfFUqsogA)

Note
- in line 10, we have a default value for `x` potentially overwritten by the spread of a `Partial`
- in line 11, we have a spread of a non-`Partial`, overwriting `y`

I'm still on the fence whether the change from 3.8.3 to 3.9.2 was positive or negative in general. For this example:
- [pro] the additional type annotations express some of the reasoning behind why the code works as it does
- [contra] typescript started marking perfectly valid code as incorrect and we had to go in and convince the compiler that the code was indeed correct as it stands
